src/riak_kv_vnode.erl

```diff
@@ -2322,102 +2346,286 @@ new_md_cache(VId) ->
+%% asked for a new lease, ask for one.  If we're less than 80% through
+%% the lease, do nothing.  If not yet blocking(equal) and we already
+%% asked for a new lease, do nothing.
+maybe_lease_counter(State=#state{counter=#counter_state{cnt=Lease, lease=Lease,
+                                                        leasing=true}}) ->
+    %% Block until we get a new lease, or crash the vnode
+    {ok, NewState} = blocking_lease_counter(State),
+    NewState;
+maybe_lease_counter(State=#state{counter=#counter_state{leasing=true}}) ->
+    %% not yet at the blocking stage, waiting on a lease
+    State;
+maybe_lease_counter(State) ->
+    #state{status_mgr_pid=MgrPid, counter=CS=#counter_state{cnt=Cnt, lease=Lease,
+                                                            lease_size=LeaseSize}} = State,
+    %% @TODO (rdb) configurable??
+    %% has more than 80% of the lease been used?
+    CS2 = if (Lease - Cnt) =< 0.2 * LeaseSize  ->
+                  ok = riak_kv_vnode_status_mgr:lease_counter(MgrPid, LeaseSize),
+                  CS#counter_state{leasing=true};
+             ?ELSE ->
+                  CS
+          end,
+    State#state{counter=CS2}.
+
+%% @private by now, we have to be waiting for a lease, or an exit,
+%% from the mgr. Block until we receive a lease. If we get an exit,
+%% retry a number of times. If we get neither in a reasonable time,
+%% return an error.
+-spec blocking_lease_counter(#state{}) ->
+                                    {ok, #state{}} |
+                                    counter_lease_error().
+blocking_lease_counter(State) ->
+    {MaxErrs, MaxTime} = get_counter_wait_values(),
+    blocking_lease_counter(State, {0, MaxErrs, MaxTime}).
+
+-spec blocking_lease_counter(#state{}, {Errors :: non_neg_integer(),
+                                        MaxErrors :: non_neg_integer(),
+                                        TimeRemainingMillis :: non_neg_integer()}
+                            ) ->
+                                    {ok, #state{}} |
+                                    counter_lease_error().
+blocking_lease_counter(_State, {MaxErrs, MaxErrs, _MaxTime}) ->
+    {error, counter_lease_max_errors};
+blocking_lease_counter(State, {ErrCnt, MaxErrors, MaxTime}) ->
+    #state{idx=Index, vnodeid=VId, status_mgr_pid=Pid, counter=CounterState} = State,
+    #counter_state{lease_size=LeaseSize} = CounterState,
+    Start = os:timestamp(),
+    receive
+        {'EXIT', Pid, Reason} ->
+            lager:error("Failed to lease counter ~p", [Reason]),
+            {ok, NewPid} = riak_kv_vnode_status_mgr:start_link(self(), Index),
+            ok = riak_kv_vnode_status_mgr:lease_counter(NewPid, LeaseSize),
+            NewState = State#state{status_mgr_pid=NewPid},
+            Elapsed = timer:now_diff(os:timestamp(), Start),
+            blocking_lease_counter(NewState, {ErrCnt+1, MaxErrors, MaxTime - Elapsed});
+        {counter_lease, {Pid, VId, NewLease}} ->
```

Comment:
What happens if this message comes in late? It seems like next call to `blocking_lease_counter/3` will pull the old message of the mailbox. Is that harmful?